content/articles/implementing-preloading-and-buffering-videos-in-android-with-exoplayer/index.md

```diff
@@ -0,0 +1,277 @@
+
+> In your Manifest file, add Internet permission
+
+### Step 3 - Creating a User Interface
+In `activity_main.xml` Design a simple layout that will contain a `PlayerView`
+
+```Xml
+<com.google.android.exoplayer2.ui.PlayerView
+    android:id="@+id/player_view"
+    android:layout_width="0dp"
+    android:layout_height="0dp"
+    app:layout_constraintBottom_toBottomOf="parent"
+    app:layout_constraintEnd_toEndOf="parent"
+    app:layout_constraintHorizontal_bias="0.5"
+    app:layout_constraintStart_toStartOf="parent"
+    app:layout_constraintTop_toTopOf="parent"
+    app:show_buffering="when_playing"
+    app:show_shuffle_button="true" />
+```
+
+![Demo2](section-engineering/implementing-preloading-and-buffering-videos-in-android-with-exoplayer/demo2.png)
+
+
+
+
+#### Explanation
+We have defined the cache size that our app will use. Also, we have defined the cache evictor which clears our cache,`ExoDatabaseProvider`, and passed all of them inside our `cache` instance.
+
+
+### Step 5 - Creating a Video Preloading Worker
+Here, we create a Worker class from the Workmanager library that will do the preloading and precaching work in the background.
+
+```Kotlin
+class VideoPreloadWorker(private val context: Context, workerParameters: WorkerParameters) :
+    Worker(context, workerParameters) {
+
+    private var videoCachingJob: Job? = null
+    private lateinit var mHttpDataSourceFactory: HttpDataSource.Factory
+    private lateinit var mDefaultDataSourceFactory: DefaultDataSourceFactory
+    private lateinit var mCacheDataSource: CacheDataSource
+    private val cache: SimpleCache = VideoApp.cache
+    
+    ...
+}        
+```
+
+Inside the class, define a companion object that will contain a method for receiving arguments from where the Worker class will be instantiated.
+
+```Kotlin
+companion object {
+    const val VIDEO_URL = "video_url"
+        
+    fun buildWorkRequest(yourParameter: String): OneTimeWorkRequest {
+        val data = Data.Builder().putString(VIDEO_URL, yourParameter).build()
+        return OneTimeWorkRequestBuilder<VideoPreloadWorker>().apply { setInputData(data) }
+        .build()
+    }
+}
+```
+
+For the video caching logic, let's define two methods that will make that work
+
+```Kotlin
+private fun preCacheVideo(videoUrl: String?) {
+
+    val videoUri = Uri.parse(videoUrl)
+    val dataSpec = DataSpec(videoUri)
+
+    val progressListener = CacheWriter.ProgressListener { requestLength, bytesCached, _ ->
+        val downloadPercentage: Double = (bytesCached * 100.0 / requestLength)
+        // Do Something
+    }
+
+    videoCachingJob = GlobalScope.launch(Dispatchers.IO) { 
+        cacheVideo(dataSpec, progressListener)
+        preCacheVideo(videoUrl)
+    } 
+}
+
+private fun cacheVideo(mDataSpec: DataSpec, mProgressListener: CacheWriter.ProgressListener) {
+    runCatching {
+        CacheWriter(
+            mCacheDataSource,
+            mDataSpec,
+            null,
+            mProgressListener,
```

Comment:
This could fit in a single line! Kindly fix that.
Also, remember to indent your code properly and indicate the language identifier!